contributors/design-proposals/watchable-discovery.md

```diff
@@ -0,0 +1,231 @@
+type GroupList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// List of group.
+	Items []Group `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+```
+
+## Behavior
+
+Each API server in the cluster serves its own `discovery.k8s.io` API
+group. By default, this is backed by normal etcd storage, i.e. the
+semantics of the `ResourceVersion` is the one of the underlaying etcd
+storage backend.
+
+Each API server has its own `DiscoveryUpdateController` that's
+regularly reconciles the storage state of the discovery API group with
+its internal registered/announce API group registry. Moreover, when
+a new API groups is installed or removed in its HTTP handler, the
+corresponding API group object is updated synchronously.
+
+### Aggegration
+
+The aggregator watches the discovery API group of each of its proxied
+API servers. It filters the returned discovery information by taking
+the `APIService` status into consideration, i.e. by 
+- only showing those GroupVersions that are registered via an
+  `APIService`
+- only showing those GroupVersions that are accepted by the aggregator
+  (i.e. which have no naming conflict).
+
+The aggregator will update its own discovery API group resources
+triggered by changes in the sub-API-servers.
+
+### Embedded Aggregator in kube-apiserver
+
+In this mode, both the aggregator and the kube-apiserver share the
+storage. To avoid conflicts about updating the API group, the `Group`
+objects of kube-apiserver are marked as "local" with an
+annotation. The aggregator will not touch them and the other way
+around for the kube-apiserver for the not-marked `Group` objects.
+
+### HA
+
+API servers in HA mode will share a storage. Hence, they will share
+the state of the discovery API group. Their controllers will fight for
+the correct state of the discovery information.
+
+During normal operation the desired discovery state is equal in an HA
+setup. So no real race will take place.
+
+During rolling updates of the control plane, an old and a new API
+server version might be active. Their reconciliation loops will race
+to update the stored groups.
+
+The consumers of the discovery information might notice the
+flipping. But, this is not a regression of the cluster behaviour as it
+already happens now if a load-balancer distributes requests among the
+cluster API servers, each returning different discovery information.
+
+A long reconciliation period will reduce the effect of the flipping,
+e.g. an API server update shouldn't take longer than a minute.
+
+We could introduce a server version for each API group. Then an older
+server could be less aggressive about reconciliation of new
+groups. But, this is not part of the API types yet.
+
+### Alternative Architectures
+
+- We could implement the discovery resource in the aggregator only and
+  use polling to the sub-API-servers. This would look mostly the same
```

Comment:
Given how complex this proposal is, I don't think you've adequately justified why simple polling is worse than big-complex-new-thing.